File: projects/grind-75/python/tree/balanced_bt.py
from typing import Optional
from tree.common import TreeNode


class Solution:
    @classmethod
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        isBalanced = True

        def dfs(node=root):
            nonlocal isBalanced
            if not node:
                return 0
            max_height_l = dfs(node.left)
            max_height_r = dfs(node.right)
            if abs(max_height_l - max_height_r) > 1:
                isBalanced = False
            return max(max_height_l, max_height_r) + 1

        dfs()
        return isBalanced

File: projects/grind-75/typescript/tree/balanced-bt.ts
import { TreeNode } from './common';

export default function isBalanced(root: TreeNode | null): boolean {
  let isBalanced = true;

  function dfs(node: TreeNode | null): number {
    if (!node) {
      return 0;
    }
    const maxHeightL = dfs(node.left);
    const maxHeightR = dfs(node.right);
    if (Math.abs(maxHeightL - maxHeightR) > 1) {
      isBalanced = false;
    }
    return Math.max(maxHeightL, maxHeightR) + 1;
  }

  dfs(root);

  return isBalanced;
}
